src/codegen/wasm.zig

```diff
@@ -305,53 +314,54 @@ pub const Context = struct {
         return WValue{ .local = self.local_index };
     }
 
-    fn genAdd(self: *Context, inst: *Inst.BinOp) InnerError!WValue {
+    fn genBinOp(self: *Context, inst: *Inst.BinOp, op: wasm.Op) InnerError!WValue {
         const lhs = self.resolveInst(inst.lhs);
         const rhs = self.resolveInst(inst.rhs);
 
         try self.emitWValue(lhs);
         try self.emitWValue(rhs);
 
-        const opcode: wasm.Opcode = switch (inst.base.ty.tag()) {
-            .u32, .i32 => .i32_add,
-            .u64, .i64 => .i64_add,
-            .f32 => .f32_add,
-            .f64 => .f64_add,
-            else => return self.fail(inst.base.src, "TODO - Implement wasm genAdd for type '{s}'", .{inst.base.ty.tag()}),
-        };
-
-        try self.code.append(wasm.opcode(opcode));
-        return .none;
+        switch (inst.base.ty.tag()) {
+            .u32, .i32, .u64, .i64, .f32, .f64 => {
+                const opcode: wasm.Opcode = wasm.buildOpcode(.{
+                    .op = op,
+                    .valtype1 = try self.typeToValtype(inst.base.src, inst.base.ty),
+                });
+                try self.code.append(wasm.opcode(opcode));
+                return .none;
+            },
+            else => return self.fail(inst.base.src, "TODO - Implement wasm genBinOp for type '{s}'", .{inst.base.ty.tag()}),
+        }
     }
 
     fn emitConstant(self: *Context, inst: *Inst.Constant) InnerError!void {
         const writer = self.code.writer();
         switch (inst.base.ty.tag()) {
-            .u32 => {
-                try writer.writeByte(wasm.opcode(.i32_const));
-                try leb.writeILEB128(writer, inst.val.toUnsignedInt());
-            },
-            .i32, .bool => {
-                try writer.writeByte(wasm.opcode(.i32_const));
-                try leb.writeILEB128(writer, inst.val.toSignedInt());
-            },
-            .u64 => {
-                try writer.writeByte(wasm.opcode(.i64_const));
-                try leb.writeILEB128(writer, inst.val.toUnsignedInt());
-            },
-            .i64 => {
-                try writer.writeByte(wasm.opcode(.i64_const));
-                try leb.writeILEB128(writer, inst.val.toSignedInt());
-            },
-            .f32 => {
-                try writer.writeByte(wasm.opcode(.f32_const));
-                // TODO: enforce LE byte order
-                try writer.writeAll(mem.asBytes(&inst.val.toFloat(f32)));
-            },
-            .f64 => {
-                try writer.writeByte(wasm.opcode(.f64_const));
-                // TODO: enforce LE byte order
-                try writer.writeAll(mem.asBytes(&inst.val.toFloat(f64)));
+            .u32, .u64, .i32, .i64, .bool, .f32, .f64 => |tag| {
```

Comment:
Perhaps we should pre-check for `void` and exit early. Then you can remove the outer switch and only leave the inner switch intact to write the correct bytes.

Reply:
The outer switch here should still exist for constants that need to be compiled to structs (or multi values), e.g. `null: ?i64`, because emitting these constants requires more than 1 wasm op, right? I can make this change for now and worry about the struct thing later as well, if you want.

Reply:
Good point! Perhaps we should be switching on a different type/tag instead? I.e. switching on zigTypeTag() would allow us to switch on .Int, .Pointer etc. and means we don't have to be as specific in regards to the tags for now.

Reply:
Done.